File: leetcode/binaryTree/1161.cpp
#include <algorithm>
#include <functional> //包含std::greater 和 std::less
#include <iostream>
#include <limits.h>
#include <map>
#include <math.h>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
class Solution {
public:
  int maxLevelSum(TreeNode *root) {
    queue<TreeNode *> treeNodes;
    treeNodes.push(root);
    int max_num = INT32_MIN;
    int max_row_id = -1;
    int cur_row_id = 1;
    while (!treeNodes.empty()) {
      // n 这一层的节点数量
      int n = treeNodes.size();

      int ans = 0;
      for (int i = 0; i < n; i++) {
        ans += treeNodes.front()->val;
        if (nullptr != treeNodes.front()->left) {
          treeNodes.push(treeNodes.front()->left);
        }
        if (nullptr != treeNodes.front()->right) {
          treeNodes.push(treeNodes.front()->right);
        }
        treeNodes.pop();
      }
      if (ans > max_num) {
        max_num = ans;
        max_row_id = cur_row_id;
      }
      cur_row_id++;
    }
    return max_row_id;
  }
};
int main() {
  // 示例二叉树
  Solution solution;
  TreeNode *root = new TreeNode(1);
  root->left = new TreeNode(7);
  root->right = new TreeNode(0);
  root->left->left = new TreeNode(7);
  root->left->right = new TreeNode(-8);
  int result = solution.maxLevelSum(root);
  std::cout << "result: " << result << std::endl;
}

File: leetcode/binaryTree/1161.py
from typing import List, Optional
from queue import Queue
from functools import cache
from collections import defaultdict


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        treeNodes = [root]
        max_num = float('-inf')
        max_row_id = -1
        cur_row_id = 1
        while treeNodes:
            tmp = treeNodes
            treeNodes = []
            ans = 0
            for node in tmp:
                ans += node.val
                if node.left:
                    treeNodes.append(node.left)
                if node.right:
                    treeNodes.append(node.right)
            if ans > max_num:
                max_num = ans
                max_row_id = cur_row_id
            cur_row_id+=1
        return max_row_id


if __name__ == "__main__":
    solution = Solution()
    root = TreeNode(1)
    root.left = TreeNode(7)
    root.right = TreeNode(0)
    root.left.left = TreeNode(7)
    root.left.right = TreeNode(-8)
    result = solution.maxLevelSum(root)
    print(f"result = {result}")
